### ISSUE TITLE
Distance Expense - Impossible to remove a waypoint when editing a distance request offline

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#34686]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.26-1
**Reproducible in staging?:** Yes 
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @paultsimura
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1705505264249909 & https://expensify.slack.com/archives/C01GTK53T8Q/p1705484273717249

## Action Performed:
1. Create a distance request with 3 waypoints
2. Navigate to the distance request details
3. Disable the internet connection
4. Click on the distance field to edit
5. Click on a waypoint
6. Click on the 3 dot menu > Delete waypoint
7. Hit Save
8. Go back online

## Expected Result:
The waypoint should be deleted. 

## Actual Result:
The waypoint is not deleted after hitting the option to delete. After going back online, the 3 waypoints are still displayed. 

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/44479856/11060808-8073-4d2a-b140-af19a804c88b

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by isagoico** [comment]
Posting @paultsimura [proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1705484389992059?thread_ts=1705484273.717249&cid=C01GTK53T8Q) here: 

## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When editing a Distance request offline, the waypoint cannot be removed
### What is the root cause of that problem?
This is because we do not handle the Editing flow correctly when passing the isDraft parameter to `Transaction.removeWaypoint`: we pass `true` always:
https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L145
https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L165
When we are creating a new request, we correctly remove the waypoint from a draft transaction. But when we edit an existing transaction, we should modify the real, not draft transaction.
### What changes do you think we should make in order to solve the problem?
We should pass the `isDraft` parameter conditionally depending on the IOU action, similar to how we pass it to `Transaction.saveWaypoint`:
```action === CONST.IOU.ACTION.CREATE```
https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L138
### What alternative solutions did you explore? (Optional)

**Comment by paultsimura** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When editing a Distance request offline, the waypoint cannot be removed
### What is the root cause of that problem?
This is because we do not handle the Editing flow correctly when passing the isDraft parameter to `Transaction.removeWaypoint`: we pass `true` always:

https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L145

https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L165

When we are creating a new request, we correctly remove the waypoint from a draft transaction. But when we edit an existing transaction, we should modify the real, not draft transaction.
### What changes do you think we should make in order to solve the problem?
We should pass the `isDraft` parameter conditionally depending on the IOU action, similar to how we pass it to `Transaction.saveWaypoint`:
```action === CONST.IOU.ACTION.CREATE```

https://github.com/Expensify/App/blob/1f72b5f1ae31765183a2a7f51a450f1e9e27b1ff/src/pages/iou/request/step/IOURequestStepWaypoint.js#L138

### What alternative solutions did you explore? (Optional)

**Comment by isagoico** [comment]
@tgolen confirmed the bug and the proposed fix here: https://expensify.slack.com/archives/C049HHMV9SM/p1705505264249909

**Comment by Vlad-AIMaster** [comment]
I will take a look at the issue and come up with a solution.

**Comment by lschurr** [comment]
@rushatgabhane could you take a look at this one?

**Comment by lschurr** [comment]
Bump @rushatgabhane

**Comment by alitoshmatov** [comment]
I will try to take a look at this today.

**Comment by alitoshmatov** [comment]
@paultsimura Thank you for proposal. Your RCA is correct. Your solution also works and on point.

**Comment by alitoshmatov** [comment]
@DylanDylann Thank you for proposal. Your proposal is very similar to @paultsimura 's proposal. I think it would be fair to go with his proposal.

**Comment by alitoshmatov** [comment]
We can go with @paultsimura 's [proposal](https://github.com/Expensify/App/issues/34686#issuecomment-1896980125).

C+ reviewed 🎀 👀 🎀

**Comment by lschurr** [comment]
Any update on this one? Are we waiting on a PR from @paultsimura?

**Comment by paultsimura** [comment]
> Any update on this one? Are we waiting on a PR from @paultsimura?

The PR is in the making. There is a Slack discussion ongoing: https://expensify.slack.com/archives/C01GTK53T8Q/p1706543295176979?thread_ts=1705484273.717249&cid=C01GTK53T8Q

**Comment by paultsimura** [comment]
After a [discussion in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1706543295176979?thread_ts=1705484273.717249&cid=C01GTK53T8Q), it was decided to (in addition to the FE changes) modify the `UpdateMoneyRequestDistance` API endpoint on the server to calculate the difference between existing and new waypoints, and explicitly return the removed ones as `null` in the response in the transaction's `merge` operation.

For example:
```js
{
  w0: A
  w1: B (remove)
  w2: C
  w3: D (remove)
}
```

Should turn into:
```js
{
  w0: A
  w1: C
  w2: null
  w3: null
}
```

cc: @tgolen @alitoshmatov

**Comment by lschurr** [comment]
Just clarifying @paultsimura - are we putting this issue on hold for the other? https://github.com/Expensify/App/issues/35503

**Comment by paultsimura** [comment]
> Just clarifying @paultsimura - are we putting this issue on hold for the other? #35503

Yes, please

**Comment by lschurr** [comment]
On hold.

**Comment by tgolen** [comment]
The backend fix for this has been deployed to production so I am taking this off HOLD.

**Comment by lschurr** [comment]
What's next on this one? @tgolen @blimpich @paultsimura

**Comment by paultsimura** [comment]
> What's next on this one? @tgolen @blimpich @paultsimura 

There seems to be a larger Onyx issue here, I'll update today or maximum tomorrow

**Comment by paultsimura** [comment]
Just posted an update on the PR.

**Comment by lschurr** [comment]
Where are we at on this one @blimpich @paultsimura?

**Comment by paultsimura** [comment]
@lschurr we should put this on hold for https://github.com/Expensify/App/issues/37560

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
On hold

**Comment by lschurr** [comment]
Hold

**Comment by mallenexpensify** [comment]
Posted on [the PR](https://github.com/Expensify/react-native-onyx/pull/490#issuecomment-2145716489) linked to https://github.com/Expensify/App/issues/37560

**Comment by lschurr** [comment]
On hold

**Comment by lschurr** [comment]
On hold

**Comment by alitoshmatov** [comment]
I couldn't reproduce this issue, it has been a long time and it looks like it was resolved along the way

**Comment by lschurr** [comment]
On hold

**Comment by lschurr** [comment]
On hold

**Comment by lschurr** [comment]
On hold

**Comment by lschurr** [comment]
Hold

**Comment by lschurr** [comment]
Still on hold

**Comment by alitoshmatov** [comment]
Recently had a similar case and got this [comment](https://github.com/Expensify/App/issues/44824#issuecomment-2272323533) by @mallenexpensify . Not completely sure if it also applies here

>VictoriaExpensify both nyomanjyotisa and alitoshmatov are due compensation here. Per the [internal SO](https://stackoverflowteams.com/c/expensify/questions/8590)

> If a contributor has been hired for a job and we decide to close the job before it is successfully completed, full payment is due for C+ and the contributor. One caveat here might be if the hired contributor wasn't working on their PR with urgency. If you're unsure, post in #bug-zero.

**Comment by lschurr** [comment]
Thanks @paultsimura and @alitoshmatov - I checked with the team and we agreed that we should pay full comp for the contributor on the PR and half comp to the C+ since proposals were reviewed but the PR was not.

Payment summary:
- Contributor: @paultsimura $500 ([Offer sent in Upwork](https://www.upwork.com/nx/wm/offer/103495665))
- Reviewer: @alitoshmatov $250 ([Offer sent in Upwork](https://www.upwork.com/nx/wm/offer/103495692))

**Comment by lschurr** [comment]
All set!
